File: src/main/java/fr/deroffal/aoc/Day05.java
package fr.deroffal.aoc;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

import fr.deroffal.aoc.util.Utils;

public class Day05 {

	public static void main(String[] args) {
		List<Integer> input = Utils.readLines("day05.txt").stream().mapToInt(Integer::valueOf).boxed().collect(Collectors.toList());

		System.out.println("Partie 1 : " + countStepToEscape(input, i -> 1));
		System.out.println("Partie 2 : " + countStepToEscape(input, i -> i >= 3 ? -1 : 1));
	}

	public static int countStepToEscape(final List<Integer> input, final Function<Integer, Integer> offsetFunction) {
		final List<Integer> instructions = new ArrayList<>(input);
		int count = 0;
		int position = 0;
		while (position < instructions.size()) {
			position = getNextPosition(instructions, position, offsetFunction);
			++count;
		}
		return count;
	}

	public static int getNextPosition(final List<Integer> instructions, final int currentPosition, final Function<Integer, Integer> offsetFunction) {
		int currentPositionValue = instructions.get(currentPosition);
		instructions.set(currentPosition, currentPositionValue + offsetFunction.apply(currentPositionValue));
		return currentPosition + currentPositionValue;
	}

}

File: kotlin/src/main/kotlin/fr/deroffal/aoc/kotlin/Day05.kt
package fr.deroffal.aoc.kotlin

import fr.deroffal.aoc.kotlin.tools.readLines

fun main(args: Array<String>) {
    val input = readLines("day05.txt").map { it.toInt() }.toIntArray()
    val day05 = Day05()
    println("Part 1 : ${day05.solvePart1(input.clone())}")
    println("Part 2 : ${day05.solvePart2(input.clone())}")
}

class Day05 {

    fun solvePart1(input: IntArray): Int = jump(input)

    fun solvePart2(input: IntArray): Int = jump(input) { if (it >= 3) -1 else 1 }

    private fun jump(input: IntArray, offsetFunction: (Int) -> Int = { 1 }): Int {
        var count = 0
        var position = 0
        while (position < input.size) {
            position = getNextPosition(input, position, offsetFunction)
            ++count
        }
        return count
    }

    private fun getNextPosition(instructions: IntArray, position: Int, offsetFunction: (Int) -> Int): Int {
        val currentPositionValue = instructions[position]
        instructions[position] = currentPositionValue + offsetFunction(currentPositionValue)
        return position + currentPositionValue
    }

}